src/main/com/yetanalytics/markdoc.clj

```diff
@@ -0,0 +1,113 @@
+            [clj-jgit.porcelain :refer [with-repo git-log]])
+  (:import [java.io File]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Markdown content -> HTML content
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn md-ext->html-ext
+  "Convert the `.md` extension of `file-path` to `.html`."
+  [file-path]
+  (cstr/replace file-path #"\.md" ".html"))
+
+(def relative-md-path-re
+  (re-pattern
+   (str "\\("
+        "(?!www\\.|(?:http|ftp)s?://|[A-Za-z]:\\\\|//)" ; no web URLs
+        "[A-Za-z0-9_\\-\\./]+"
+        "\\.md"
+        "(?:#[\\w\\?\\/\\\\\\:\\@\\.\\~\\!\\$\\&\\'\\(\\)\\*\\+\\,\\;\\=]*)?" ; fragments
+        "\\)")))
+
+(defn relative-md-links
+  "Convert Markdown links in `text` into HTML links"
+  [text state]
+  [(cstr/replace text relative-md-path-re md-ext->html-ext)
+   state])
+
+(defn md->html
+  "Wrapper for `md/md-to-html-string`."
+  [text]
+  (md/md-to-html-string
+   text
+   :heading-anchors true
+   :replacement-transformers (cons relative-md-links
+                                   md-trans/transformer-vector)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Git API for SHA
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Shut up VSCode warnings.
+;; Unfortunately `with-repo` is a unique macro where you have a floating `repo`
+;; symbol that is only bound after macroexpansion, so we cannot use `lint-as`
+;; to lint it like a similar macro.
+#_{:clj-kondo/ignore [:unresolved-symbol]}
+(def git-sha
+  (delay
+    (try
+      (with-repo "./"
+        (-> (git-log repo :max-count 1)
+            first
+            :id
+            (.getName)))
+      (catch Exception _ "No SHA found."))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Markdown file -> HTML file
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+
+(def get-template
+  (memoize (fn [template-file]
+             (slurp template-file))))
+
+(defn fill-template
+  "Add `content` to the HTML doc template."
+  [content template-file & {:keys [template-vars]}]
+  (selm-parser/render (get-template template-file)
+                      (merge {:content content
+                              :sha @git-sha}
+                             template-vars)))
+
+(defn all-paths-seq
+  "Return a seq of all files located in `root`."
+  [root]
+  (->> (io/file root)
+       file-seq
+       (remove #(.isDirectory ^File %))
+       (map #(.getPath ^File %))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Markdown doc folder -> HTML doc folder
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn convert
+  "Given Markdown docs at `in-root`, create HTML docs in `out-root`."
+  [{:keys [in-root out-root template-file template-vars]}]
```

Comment:
You say that `template-file` is optional but we don't provide a default?

Reply:
Like what would happen if you _don't_ provide a `template-file`? Will it break?

Reply:
And somewhat related - what if you're missing `template-vars` even if the Selmer template you're using has them? Will they just be passed in as empty strings or would it also break?

Reply:
And finally you might want to think of what to do if `in-root` and/or `out-root` are missing keys.

Reply:
No. Template-file is totally not required. Without template-file it will just directly convert the MD into HTML without a template. This supports some older use cases.

In terms of :in-root and :out-root, if the fn is called directly via import it will error, but if called from CLI it will default to the test locations, which will convert a sample MD and dump them in target. That works both ways in the sense that if you JUST give an :out it will convert and dump some samples to where you say, and if you JUST give an :in it will dump your converted md into a target/ location.

Reply:
I went ahead and put in some asserts so the errors are clearer.

Reply:
Ok - might want to make that behavior more obvious in the documentation.

Reply:
I mean, the whole lib's purpose is taking a required in and an out, and that's plainly written as required in docs. The sample dirs are a convenience, but not the use case of the app. I think you're taking this too far.